File: frontend/src/app/core/services/harm-severity.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { ApiResponse } from '../models/api-response.model';
import { PageResponse } from '../models/page.model';
import {
  HarmSeverity,
  HarmSeverityCreateRequest,
  HarmSeverityStatusUpdateRequest,
  HarmSeverityUpdateRequest,
} from '../models/harm-severity.model';

/**
 * Serviço responsável pela comunicação HTTP com os endpoints de Gravidades de Dano (NCC MERP).
 */
@Injectable({
  providedIn: 'root',
})
export class HarmSeverityService {
  private readonly http = inject(HttpClient);
  private readonly baseUrl = '/api/gtt/severities';

  /**
   * Consulta o guia interativo de gravidades ordenadas de A a I.
   */
  getGuide(): Observable<ApiResponse<HarmSeverity[]>> {
    return this.http.get<ApiResponse<HarmSeverity[]>>(`${this.baseUrl}/guide`);
  }

  /**
   * Lista gravidades com paginação e filtros administrativos.
   */
  listSeverities(
    search: string = '',
    isHarm: boolean | null = null,
    isActive: boolean | null = null,
    page: number = 0,
    size: number = 10,
    sort: string = 'categoryLetter,asc'
  ): Observable<ApiResponse<PageResponse<HarmSeverity>>> {
    let params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString())
      .set('sort', sort);

    if (search && search.trim()) {
      params = params.set('search', search.trim());
    }
    if (isHarm !== null) {
      params = params.set('isHarm', isHarm.toString());
    }
    if (isActive !== null) {
      params = params.set('isActive', isActive.toString());
    }

    return this.http.get<ApiResponse<PageResponse<HarmSeverity>>>(this.baseUrl, { params });
  }

  /**
   * Lista gravidades de dano com filtro ativo para seleção clínica.
   */
  listHarmSeverities(
    search: string = '',
    isActive: boolean | null = true,
    page: number = 0,
    size: number = 20
  ): Observable<ApiResponse<PageResponse<HarmSeverity>>> {
    return this.listSeverities(search, true, isActive, page, size);
  }

  /**
   * Obtém os dados de uma gravidade por ID.
   */
  getSeverityById(id: string): Observable<ApiResponse<HarmSeverity>> {
    return this.http.get<ApiResponse<HarmSeverity>>(`${this.baseUrl}/${id}`);
  }

  /**
   * Cadastra uma nova categoria de gravidade.
   */
  createSeverity(data: HarmSeverityCreateRequest): Observable<ApiResponse<HarmSeverity>> {
    return this.http.post<ApiResponse<HarmSeverity>>(this.baseUrl, data);
  }

  /**
   * Atualiza uma categoria de gravidade existente.
   */
  updateSeverity(id: string, data: HarmSeverityUpdateRequest): Observable<ApiResponse<HarmSeverity>> {
    return this.http.put<ApiResponse<HarmSeverity>>(`${this.baseUrl}/${id}`, data);
  }

  /**
   * Altera o status ativo/inativo de uma categoria de gravidade.
   */
  updateStatus(id: string, data: HarmSeverityStatusUpdateRequest): Observable<ApiResponse<HarmSeverity>> {
    return this.http.patch<ApiResponse<HarmSeverity>>(`${this.baseUrl}/${id}/status`, data);
  }

  /**
   * Exclui uma categoria de gravidade.
   */
  deleteSeverity(id: string): Observable<ApiResponse<void>> {
    return this.http.delete<ApiResponse<void>>(`${this.baseUrl}/${id}`);
  }
}

File: backend/src/main/java/br/ufs/sigea/gtt/severity/service/HarmSeverityService.java
package br.ufs.sigea.gtt.severity.service;

import br.ufs.sigea.common.dto.PageResponse;
import br.ufs.sigea.common.exception.BusinessException;
import br.ufs.sigea.common.exception.ResourceNotFoundException;
import br.ufs.sigea.gtt.severity.domain.HarmSeverity;
import br.ufs.sigea.gtt.severity.dto.HarmSeverityCreateDTO;
import br.ufs.sigea.gtt.severity.dto.HarmSeverityResponseDTO;
import br.ufs.sigea.gtt.severity.dto.HarmSeverityStatusUpdateDTO;
import br.ufs.sigea.gtt.severity.dto.HarmSeverityUpdateDTO;
import br.ufs.sigea.gtt.severity.mapper.HarmSeverityMapper;
import br.ufs.sigea.gtt.severity.repository.HarmSeverityRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.UUID;

/**
 * Serviço de regras de negócio para a gestão de Categorias de Gravidade de Dano NCC MERP.
 */
@Service
@RequiredArgsConstructor
@Slf4j
public class HarmSeverityService {

    private final HarmSeverityRepository severityRepository;
    private final HarmSeverityMapper severityMapper;

    /**
     * Lista categorias de gravidade paginadas com filtros por busca textual, flag de dano e status.
     *
     * @param search   Termo de pesquisa por letra, nome ou descrição
     * @param isHarm   Filtro opcional por flag de dano real
     * @param isActive Filtro opcional por status ativo
     * @param pageable Configuração de paginação
     * @return Página de categorias de gravidade
     */
    @Transactional(readOnly = true)
    public PageResponse<HarmSeverityResponseDTO> listSeverities(String search, Boolean isHarm, Boolean isActive, Pageable pageable) {
        String query = (search != null && !search.trim().isEmpty()) ? search.trim() : null;
        Page<HarmSeverity> page = severityRepository.findWithFilters(query, isHarm, isActive, pageable);
        return PageResponse.from(page.map(severityMapper::toResponseDTO));
    }

    /**
     * Retorna todas as categorias ativas ordenadas de A a I para o guia interativo educacional.
     *
     * @return Lista ordenada de categorias
     */
    @Transactional(readOnly = true)
    public List<HarmSeverityResponseDTO> getGuide() {
        return severityRepository.findAllByIsActiveTrueOrderByCategoryLetterAsc().stream()
                .map(severityMapper::toResponseDTO)
                .toList();
    }

    /**
     * Recupera os dados de uma gravidade específica por ID.
     *
     * @param id ID da gravidade
     * @return DTO da gravidade
     * @throws ResourceNotFoundException se não for encontrada
     */
    @Transactional(readOnly = true)
    public HarmSeverityResponseDTO getSeverityById(UUID id) {
        HarmSeverity severity = findSeverityOrThrow(id);
        return severityMapper.toResponseDTO(severity);
    }

    /**
     * Cadastra uma nova categoria de gravidade.
     *
     * @param dto Dados para cadastro
     * @return DTO da categoria criada
     * @throws BusinessException se a letra da categoria já estiver cadastrada
     */
    @Transactional
    public HarmSeverityResponseDTO createSeverity(HarmSeverityCreateDTO dto) {
        String normalizedLetter = dto.getCategoryLetter().trim().toUpperCase();

        if (severityRepository.existsByCategoryLetterIgnoreCase(normalizedLetter)) {
            throw new BusinessException("Já existe uma categoria de gravidade cadastrada com a letra '" + normalizedLetter + "'.");
        }

        HarmSeverity severity = severityMapper.toEntity(dto);
        severity.setCategoryLetter(normalizedLetter);
        severity.setName(dto.getName().trim());
        severity.setDescription(dto.getDescription().trim());
        severity.setIsHarm(dto.getIsHarm());

        HarmSeverity saved = severityRepository.save(severity);
        log.info("Categoria de gravidade criada: id={}, letter={}", saved.getId(), saved.getCategoryLetter());

        return severityMapper.toResponseDTO(saved);
    }

    /**
     * Atualiza uma categoria de gravidade existente.
     *
     * @param id  ID da categoria
     * @param dto Novos dados
     * @return DTO atualizado
     * @throws ResourceNotFoundException se não existir
     * @throws BusinessException         se houver colisão na letra identificadora
     */
    @Transactional
    public HarmSeverityResponseDTO updateSeverity(UUID id, HarmSeverityUpdateDTO dto) {
        HarmSeverity severity = findSeverityOrThrow(id);
        String normalizedLetter = dto.getCategoryLetter().trim().toUpperCase();

        if (severityRepository.existsByCategoryLetterIgnoreCaseAndIdNot(normalizedLetter, id)) {
            throw new BusinessException("Já existe outra categoria de gravidade com a letra '" + normalizedLetter + "'.");
        }

        severityMapper.updateEntityFromDTO(dto, severity);
        severity.setCategoryLetter(normalizedLetter);
        severity.setName(dto.getName().trim());
        severity.setDescription(dto.getDescription().trim());
        severity.setIsHarm(dto.getIsHarm());

        HarmSeverity updated = severityRepository.save(severity);
        log.info("Categoria de gravidade atualizada: id={}, letter={}", updated.getId(), updated.getCategoryLetter());

        return severityMapper.toResponseDTO(updated);
    }

    /**
     * Altera o status ativo/inativo de uma categoria de gravidade.
     *
     * @param id  ID da gravidade
     * @param dto Novo status
     * @return DTO atualizado
     */
    @Transactional
    public HarmSeverityResponseDTO updateStatus(UUID id, HarmSeverityStatusUpdateDTO dto) {
        HarmSeverity severity = findSeverityOrThrow(id);
        severity.setIsActive(dto.getIsActive());

        HarmSeverity updated = severityRepository.save(severity);
        log.info("Status da gravidade id={} alterado para isActive={}", updated.getId(), updated.getIsActive());

        return severityMapper.toResponseDTO(updated);
    }

    /**
     * Exclui uma categoria de gravidade do sistema.
     *
     * @param id ID da gravidade
     */
    @Transactional
    public void deleteSeverity(UUID id) {
        HarmSeverity severity = findSeverityOrThrow(id);
        severityRepository.delete(severity);
        log.info("Categoria de gravidade id={} excluída com sucesso", id);
    }

    /**
     * Busca uma gravidade por ID ou lança ResourceNotFoundException.
     *
     * @param id ID da gravidade
     * @return Entidade HarmSeverity
     */
    private HarmSeverity findSeverityOrThrow(UUID id) {
        return severityRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Categoria de gravidade não encontrada com o ID: " + id));
    }
}
